Improve `CellPath` display output

aionescu opened the pull request:
<!--
if this PR closes one or more issues, you can automatically link the PR with
them by using one of the [*linking keywords*](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), e.g.
- this PR should close #xxxx
- fixes #xxxx

you can also mention related issues, PRs or discussions!
-->

# Description
<!--
Thank you for improving Nushell. Please, check our [contributing guide](../CONTRIBUTING.md) and talk to the core team before making major changes.

Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.
-->

Fixes: #13362

This PR fixes the `Display` impl for `CellPath`, as laid out in #13362 and #14090:

```nushell
> $.0."0"
$.0."0"

> $."foo.bar".baz
$."foo.bar".baz
```

# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->

Cell-paths are now printed using the same `$.` notation that is used to create them, and ambiguous column names are properly quoted.

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use toolkit.nu; toolkit test stdlib"` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->

aionescu commented:
Started this as a draft because there's some details to iron out (CC @fdncred @sholderbach):

* This patch breaks the output of `select`, even in the most basic cases like `select x`. The output column is now named `$.x` (as a string).
* This patch also breaks `flatten` for some reason (hence the failing tests): It now just returns the input table unmodified. I haven't properly looked investigated yet, but at first glance `flatten` [pops a `Vec<CellPath>` of the call-stack](https://github.com/nushell/nushell/blob/main/crates/nu-command/src/filters/flatten.rs#L123), which looks like the culprit.
  * Is this a bug in `flatten`? Its signature says it accepts _string_ rest parameters, not cell-paths. Also, calling it with more than one column name results in an error, so maybe its signature should be changed to a single optional string parameter?

sholderbach commented:
> * This patch breaks the output of `select`, even in the most basic cases like `select x`. The output column is now named `$.x` (as a string).

I think we should have a separate function generating the `select` column name. So without the `$.` or `?` optional annotation as that is indicating something about the gathering operation and not really about the location specified by the path.

> * This patch also breaks `flatten` for some reason (hence the failing tests): It now just returns the input table unmodified. I haven't properly investigated yet, but at first glance `flatten` [pops a `Vec<CellPath>` of the call-stack](https://github.com/nushell/nushell/blob/main/crates/nu-command/src/filters/flatten.rs#L123), which looks like the culprit.

Here's the line where it naively tries to turn a `CellPath` into a string assuming that would yield a column name
https://github.com/nushell/nushell/blob/e104bccfb94e5572c4c138ce59121ed84456d7d0/crates/nu-command/src/filters/flatten.rs#L160 



> * `flatten`'s signature says it accepts _string_ rest parameters, not cell-paths. Also, calling it with more than one column name results in an error, so maybe its signature should be changed to a single optional string parameter?

Yeah but we have a silent conversion to `CellPath` available for `CallExt` to use. (and if taken from a value that is not parse time checked it wouldn't get caught by the `SyntaxShape` matching anyway)
https://github.com/nushell/nushell/blob/2c31b3db072f60869dbe49f3bb1943d76c07b21e/crates/nu-protocol/src/value/from_value.rs#L582-L595

I am suspicious of those coercions #14147

aionescu commented:
> I think we should have a separate function generating the select column name. So without the $. or ? optional annotation as that is indicating something about the gathering operation and not really about the location specified by the path.

Good point. I added a new method `CellPath::to_column_name` which has the same behavior as `CellPath`'s old `Display` (no `$.`, no `?`, no quoting), and replaced all calls to `CellPath::to_string` in `select` with `to_column_name`.

> let column_requested = columns.iter().find(|c| c.to_string() == column); 

Yeah, that looks naive indeed :) So basically `flatten` takes a string, coerces it to a `CellPath`, and then converts it back to a string?

For now, I just replaced that `to_string` with `to_column_name` to make the tests pass, but it feels like a hack.

Also, a side-effect of this PR is that more strings are quoted by `to nuon` when inside structures, because the same `needs_quotes` is used everywhere:

```nushell
> [...] | to nuon # old
[...]

> [...] | to nuon # new
["..."]

> "..." | to nuon # unchanged, outside structures it was quoted anyway
"..."
```

I personally like this (as it's more consistent between inside & outside structures), but I'm not sure if that's the intended output of `to nuon` (should it only quote when absolutely necessary?)

sholderbach commented on crates/nu-protocol/src/ast/cell_path.rs:
```diff
@@ -173,22 +174,46 @@ impl CellPath {
+                    s += &val.to_string();
+                }
+                PathMember::String { val, .. } => {
+                    s += val;
```
Here we still need quotes around number-like columns otherwise they will become indistinguishable from rows.

aionescu replied:
Yeah, but is that a problem in this situation? Since all the column names get crammed into a string anyway, I think it's easier for users to not quote indexes, e.g. `get 'a."0".x'` vs `get a.0.x`

sholderbach replied:
`0`- `PathMember::Int` -> row
`"0"` - `PathMember::String` -> column

It would be impossible to determine the meaning of something like 

```nushell
let confusing = [{"0": bla, "1": blub}, {"0": foo, "1": bar}]
let cellp = $.0."1"

print $cellp
$confusing | select $cellp
```

aionescu replied:
This example doesn't work (for unrelated reasons: `select` with a column after a row fails with `Select only allows row numbers for rows`), but I see your point.

We could fix this in a separate PR, but it would be a breaking change (similar to adding `?`, cf. #14233), so maybe @fdncred should chime in.

fdncred replied:
I'm not following? Rows should be ints and Columns should be strings. Are we talking about changing that? If so, in what way?

aionescu replied:
> Seems like `$confusing | get $cellp` works but `$confusing | select $cellp` doesn't.

That's because `select` doesn't let you mix rows and columns (see [this check](https://github.com/nushell/nushell/blob/main/crates/nu-command/src/filters/select.rs#L187)).

And I think this actually makes the point raised by @sholderbach not a problem, because the column names produced by `select` will _never_ contain row indices. e.g. if `select` produces a column `a.0.b`, that `0` came from a pre-existing column name, so there's no point quoting it?

fdncred replied:
Just my opinion, but I like the way `get` works, as far as following cell paths, better than how select works right now. I like the flexibility of being able to do `get 0.a` or `get a.0` and getting the same result.

sholderbach commented on crates/nu-command/src/filters/flatten.rs:
```diff
@@ -157,7 +157,7 @@ fn flat_value(columns: &[CellPath], item: Value, all: bool) -> Vec<Value> {
 
             for (column_index, (column, value)) in val.into_owned().into_iter().enumerate() {
-                let column_requested = columns.iter().find(|c| c.to_string() == column);
+                let column_requested = columns.iter().find(|c| c.to_column_name() == column);
```
Is `flatten` accepting a nested cell path semantically or only a single column level?

aionescu replied:
It only accepts a single (string) column name, and trying to pass it a cell-path results in an error.

sholderbach replied:
Ah OK, otherwise we could have moved away from using `CellPath` internally to something more explicit as just a string.

sholderbach commented on tests/repl/test_spread.rs:
```diff
@@ -30,7 +30,7 @@ fn not_spread() -> TestResult {
     run_test(
         r#"let a = 4; [... $a ... [1] ... (5) ...bare ...] | to nuon"#,
-        "[..., 4, ..., [1], ..., 5, ...bare, ...]",
+        r#"["...", 4, "...", [1], "...", 5, "...bare", "..."]"#,
```
That `run_test` kind of escapes the meaning of that old test and is more one for the `nuon` sanity checks right?

aionescu replied:
I think so, yes.

sholderbach reviewed (commented):
Thanks for implementing that!

Not sure how the internals of `flatten` actually behave without digging into it.

Some folks want(ed) to make NUON way more parsimonous with characters than it is sane to read ^^

I personally would be fine with us disallowing the `.` for the existence of ranges and the spread operator alone.

fdncred commented:
It would be nice to land this soon, if it's ready, since we have a breaking change in #14042 with a revert to unbreak it in #14233.

aionescu commented:
IMO it's ready, but we should wait for @sholderbach's opinion.

sholderbach commented:
We can land this and rework the last bug source https://github.com/nushell/nushell/pull/14197#discussion_r1826586599 later if we don't want to entertain the breakage during development.

fdncred commented:
ok, let's do that. i'll land this and close my revert. thanks!

amtoine commented:
marking this as a "_breaking change_" as it broke the CI of Nupm

aionescu commented:
@amtoine
How difficult is the fix?
I'd rather not revert this (as my scripts now rely on the new behavior)

amtoine commented:
i have been distracted by some other stuff tonight but a fix for Nupm is in preparation, hopefully it's not too bad, should just be a matter of getting the new hashes :wink: 

and it was not to suggest reverting this PR, no worries :yum:
